python/spear/env.py

```diff
@@ -553,3 +553,96 @@ def _serialize_array(array, space, byte_order):
+        self._wheel_rotation_speeds = obs["encoder"]
+
+        return obs, reward, is_done, step_info
+
+    def reset(self, reset_info=None):
+    
+        obs = super().reset(reset_info=reset_info)
+        assert "encoder" in obs.keys()
+        self._wheel_rotation_speeds = obs["encoder"]
+
+        return obs
+
+    def _apply_action(self, action):
+
+        # calculate wheel torques based on input PWM signals
+        if "apply_voltage" in action.keys():
+
+            assert self._wheel_rotation_speeds is not None
+            assert len(action["apply_voltage"]) == 2
+
+            duty_cycle = np.array([action["apply_voltage"][0], action["apply_voltage"][1], action["apply_voltage"][0], action["apply_voltage"][1]], dtype=np.float64)
+
+            # Motor torque: 1200 gf.cm (gram force centimeter) == 0.1177 N.m
+            # Gear ratio: 50
+            # Max wheel torque: 5.88399 N.m
+            # https://www.conrad.de/de/p/joy-it-com-motor01-getriebemotor-gelb-schwarz-passend-fuer-einplatinen-computer-arduino-banana-pi-cubieboard-raspbe-1573543.html
+
+            motor_velocity_constant = self._config.OPENBOT.MOTOR_VELOCITY_CONSTANT # Motor torque constant in [N.m/A]
+            gear_ratio              = self._config.OPENBOT.GEAR_RATIO              # Gear ratio of the OpenBot motors
+            motor_torque_constant   = 1.0 / motor_velocity_constant;                # Motor torque constant in [rad/s/V]
+
+            battery_voltage         = self._config.OPENBOT.BATTERY_VOLTAGE         # Voltage of the battery powering the OpenBot [V]
+            control_dead_zone       = self._config.OPENBOT.CONTROL_DEAD_ZONE       # Absolute duty cycle (in the ACTION_SCALE range) below which a command does not produce any torque on the vehicle
+            motor_torque_max        = self._config.OPENBOT.MOTOR_TORQUE_MAX        # Motor maximal torque [N.m]
+            electrical_resistance   = self._config.OPENBOT.ELECTRICAL_RESISTANCE   # Motor winding electrical resistance [Ohms]
+
+            # Speed multiplier defined in the OpenBot to map a [-1,1] action to a suitable command to 
+            # be processed by the low-level microcontroller. For more details, feel free to check the
+            # "speedMultiplier" command in the OpenBot code.
+            # https://github.com/isl-org/OpenBot/blob/master/android/app/src/main/java/org/openbot/vehicle/Vehicle.java#L375
+            action_scale = self._config.OPENBOT.ACTION_SCALE
+
+            # Acquire the ground truth motor and wheel velocity for motor counter-electromotive force
+            # computation purposes (or alternatively friction computation purposes).
+
+            # The ground truth velocity of the robot wheels in [RPM]
+            wheel_rotation_speeds = self._wheel_rotation_speeds
+
+            # The ground truth rotation speed of the motors in [rad/s]
+            motor_speed = gear_ratio * wheel_rotation_speeds * np.pi / 30.0 # Expressed in [rad/s]
```

Comment:
wheel_rotation_speeds is already in rad/s so it doesn't need to be converted (https://github.com/EpicGames/UnrealEngine/blob/5.2/Engine/Source/Runtime/Experimental/ChaosVehicles/ChaosVehiclesCore/Public/WheelSystem.h#L301)

Reply:
Corrected it